DataFormats/HGCDigi/interface/PHGCSimAccumulator.h

```diff
@@ -0,0 +1,156 @@
+#ifndef DataFormats_HGCDigi_PHGCSimAccumulator_h
+#define DataFormats_HGCDigi_PHGCSimAccumulator_h
+
+#include "DataFormats/DetId/interface/DetId.h"
+
+#include <vector>
+#include <cassert>
+
+class PHGCSimAccumulator {
+public:
+  // These two structs are public only because of dictionary generation
+  class DetIdSize {
+  public:
+    // Use top 27 bits to hold the details of the DetId
+    constexpr static unsigned detIdOffset = 5;
+    constexpr static unsigned detIdMask = 0x7ffffff;
+    // Use the last 5 bits to index 2x15 elements
+    constexpr static unsigned sizeOffset = 27;
+    constexpr static unsigned sizeMask = 0x1f;
+    // With this arrangement increasing the size component is faster
+
+    DetIdSize(): detIdSize_(0) {}
+    DetIdSize(unsigned int detId): detIdSize_(detId << detIdOffset) {}
+
+    void increaseSize() {
+      assert(size()+1 < 1<<detIdOffset);
+      ++detIdSize_;
+    }
+
+    unsigned int detIdDetails() const { return detIdSize_ >> detIdOffset; }
+    unsigned int size() const { return detIdSize_ & sizeMask; }
+
+  private:
+    unsigned int detIdSize_;
+  };
+  class Data {
+  public:
+    constexpr static unsigned energyOffset = 15;
+    constexpr static unsigned energyMask = 0x1;
+    constexpr static unsigned sampleOffset = 11;
+    constexpr static unsigned sampleMask = 0xf;
+    constexpr static unsigned dataOffset = 0;
+    constexpr static unsigned dataMask = 0x7ff;
+
+    Data(): data_(0) {}
+    Data(unsigned short ei, unsigned short si, unsigned short d):
+      data_((ei << energyOffset) | (si << sampleOffset) | d)
+    {}
+
+    unsigned int energyIndex() const { return data_ >> energyOffset; }
+    unsigned int sampleIndex() const { return (data_ >> sampleOffset) & sampleMask; }
+    unsigned int data() const { return data_ & dataMask; }
+
+  private:
+    unsigned short data_;
+  };
+
+  PHGCSimAccumulator() = default;
+  PHGCSimAccumulator(unsigned int detId): detSubdetId_(detId >> DetId::kSubdetOffset) {}
+  ~PHGCSimAccumulator() = default;
+
+  void reserve(size_t size) {
+    detIdSize_.reserve(size);
+    data_.reserve(size);
+  }
+
+  void shrink_to_fit() {
+    detIdSize_.shrink_to_fit();
+    data_.shrink_to_fit();
+  }
+
+  void emplace_back(unsigned int detId, unsigned short energyIndex, unsigned short sampleIndex, unsigned short data) {
+    assert( (detId >> DetId::kSubdetOffset) == detSubdetId_ );
+
+    // TODO: add checks
```

Comment:
is this todo now satisfied with the checks below? if not, what else should be added?

Reply:
I must admit I don't remember anymore exactly what I had in mind (so better write better comments next time...). Most likely the "checks" refer checkign that `energyIndex` and `sampleIndex` are small-enough to fit in the space reserved for them. On the other hand, these checks are already done in `HGCDigitizer:saveSimHitAccumulator` (with `static_assert`), so in practice there is no need to repeat them here. I could add a comment that it is the caller's responsibility to guarantee `energyIndex` and `sampleIndex` to fit.

Reply:
I think it would be useful to add such a comment.